protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,545 @@
+    ///
+    /// More specifically index into the `ClosestDisjointPeersIter::iters` vector. On the one hand
+    /// this is used to link responses from remote peers back to the corresponding iterator, on the
+    /// other hand it is used to track which peers have been contacted in the past.
+    yielded_peers: HashMap<PeerId, usize>,
+    /// Index of the iterator last queried.
+    last_quiered: usize,
+}
+
+impl ClosestDisjointPeersIter {
+    /// Creates a new iterator with a default configuration.
+    pub fn new<I>(target: KeyBytes, known_closest_peers: I) -> Self
+    where
+        I: IntoIterator<Item = Key<PeerId>>,
+    {
+        Self::with_config(
+            ClosestPeersIterConfig::default(),
+            target,
+            known_closest_peers,
+        )
+    }
+
+    /// Creates a new iterator with the given configuration.
+    pub fn with_config<I, T>(
+        config: ClosestPeersIterConfig,
+        target: T,
+        known_closest_peers: I,
+    ) -> Self
+    where
+        I: IntoIterator<Item = Key<PeerId>>,
+        T: Into<KeyBytes> + Clone,
+    {
+        let peers = known_closest_peers.into_iter().take(K_VALUE.get()).collect::<Vec<_>>();
+        let iters = split_num_results_per_disjoint_path(&config)
+            .into_iter()
+            // NOTE: All [`ClosestPeersIter`] share the same set of peers at
+            // initialization. The [`ClosestDisjointPeersIter.yielded_peers`]
+            // ensures a peer is only ever queried by a single
+            // [`ClosestPeersIter`].
+            .map(|config| ClosestPeersIter::with_config(config, target.clone(), peers.clone()))
+            .collect::<Vec<_>>();
+
+        let iters_len = iters.len();
+
+        ClosestDisjointPeersIter {
+            iters,
+            yielded_peers: HashMap::new(),
+            // Wraps around, thus iterator 0 will be queried first.
+            last_quiered: iters_len - 1,
+        }
+    }
+
+    pub fn on_failure(&mut self, peer: &PeerId) {
+        // All peer failures are reported to all queries and thus to all peer
+        // iterators. If this iterator never started a request to the given peer
+        // ignore the failure.
+        if let Some(index) = self.yielded_peers.get(peer) {
+            self.iters[*index].on_failure(peer)
+        }
+    }
+
+    pub fn on_success<I>(&mut self, peer: &PeerId, closer_peers: I)
+    where
+        I: IntoIterator<Item = PeerId>,
+    {
+        self.iters[self.yielded_peers[peer]].on_success(peer, closer_peers);
+    }
+
+    pub fn is_waiting(&self, peer: &PeerId) -> bool {
+        self.iters[self.yielded_peers[peer]].is_waiting(peer)
+    }
+
+    pub fn next(&mut self, now: Instant) -> PeersIterState {
+        let mut state = PeersIterState::Finished;
+
+        // Order in which to query the iterators to ensure fairness. Make sure
+        // to query the previously queried iterator last.
+        let iter_order = {
```

Comment:
Problem with `Cycle` is:

- It requires the `Iterator` to implement `Clone`.

- It can not take ownership of the `ClosestPeersIter`s, because one still needs to access them individually e.g. in `on_success`.

- Not taking ownership over the `ClosestPeersIter` but referencing to the `ClosestDisjointPeersIter::iters` `Vec` would require self referential structs.

> It would be good to avoid maintaining indices like last_queried.

Agree, not having to keep track of this would be great, but I have not been able to come up with a simple solution.

Reply:
I didn't really mean to imply that `Cycle` necessarily iterates though the iterators, but why not use it to iterate through the indices, like you're doing here with the `iter_order` vector together with `last_queried`? The way in which a new `Vec` is allocated here on every call to `next` that is "wrapped around" at `last_queried` just seems very cumbersome and I can't stop thinking that there must be a simpler, and more efficient approach.